pkg/cloud/azure/actuators/machine/reconciler.go

```diff
@@ -685,3 +696,36 @@ func getSpotVMOptions(spotVMOptions *v1beta1.SpotVMOptions) (compute.VirtualMach
 	// https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/spot-instances.md#eviction-policies
 	return compute.Spot, compute.Deallocate, billingProfile, nil
 }
+
+func (s *Reconciler) createAvailabilitySet() (string, error) {
+	// Try to find the availability set for the machine location
+	availabilityZones, err := s.availabilityZonesSvc.Get(context.Background(), &availabilityzones.Spec{
+		VMSize: s.scope.MachineConfig.VMSize,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	availabilityZonesSlice, ok := availabilityZones.([]string)
+	if !ok {
+		return "", fmt.Errorf("unexpected type %T", availabilityZones)
+	}
+
+	if len(availabilityZonesSlice) != 0 {
+		klog.V(4).Infof("No availability set needed for %s because availability zones were found", s.scope.Machine.Name)
+		return "", nil
+	}
+
+	klog.V(4).Infof("No availability zones were found for %s, an availability set will be created", s.scope.Machine.Name)
+
+	// Create the availability set
+	availabilityZoneName := fmt.Sprintf("%s-%s", s.scope.Machine.Labels[MachineSetLabelName], "worker")
+
+	if err := s.availabilitySetsSvc.CreateOrUpdate(context.Background(), availabilitysets.Spec{
+		Name: availabilityZoneName,
```

Comment:
Are there any other things we might want to put in here, for example does it need to be created in the correct resource group? Tags etc?

Reply:
I'm not sure about tags, but it's created in a proper resource group https://github.com/openshift/cluster-api-provider-azure/pull/225/files#diff-8d20fa6ead6f3242c2a44da1633ec1c6deaeb1f41d0dac117e812d04bc39d8e9R38